agent/consul/controller/controller_gateways.go

```diff
@@ -0,0 +1,58 @@
+package controller
+
+import (
+	"context"
+	"github.com/hashicorp/consul/agent/consul/stream"
+	"sync"
+	"time"
+)
+
+//TODO stub work this will likely need to be updated but for now it should be alright just to call to the basic implementation
+type APIGatewayController struct {
+	controller Controller
+}
+
+//TODO do we need type specific reconcilers or can we reuse them?
+type apiGatewayReconciler struct {
+	mutex sync.Mutex
+}
+
+func (r *apiGatewayReconciler) Reconcile(ctx context.Context, req Request) error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	return nil
+}
+
+func newAPIGatewayReconciler() *apiGatewayReconciler {
+	return &apiGatewayReconciler{}
+}
+
+func NewAPIGatewayController(ctx context.Context) *APIGatewayController {
```

Comment:
So, we shouldn't need any of these structs -- it should just be a matter of creating a struct that implements "Reconciler" and configuring a controller that leverages it accordingly, very similar to any of the existing k8s gateway stuff:

https://github.com/hashicorp/consul-api-gateway/blob/main/internal/k8s/controllers/gateway_class_controller.go

Really all you'll need is a method similar to the bits in `SetupWithManager` that does something like what's already in the unit tests:

https://github.com/hashicorp/consul/blob/831680d2c5b3c34caffee89168025b16f897f151/agent/consul/controller/controller_test.go#L47-L50